### PR TITLE
[Merged by Bors] - feat: characterise the reverse of the characteristic polynomial of a matrix

### PR DESCRIPTION
**Pull request by ocfnash**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Polynomial/Laurent.lean [hunk 1]
```diff
@@ -610,4 +623,31 @@ theorem isLocalization : IsLocalization (Submonoid.closure ({X} : Set R[X])) R[T
 
 end CommSemiring
 
+section Inversion
+
+variable {R : Type _} [CommSemiring R]
+
+/-- The map which substitutes `T ↦ T⁻¹` into a Laurent polynomial. -/
+def invert : R[T;T⁻¹] ≃ₐ[R] R[T;T⁻¹] := AddMonoidalgebra.domCongr k k <| AddEquiv.neg G
```
Comment (eric-wieser) [inline Mathlib/Data/Polynomial/Laurent.lean:631]:
```suggestion
def invert : R[T;T⁻¹] ≃ₐ[R] R[T;T⁻¹] := AddMonoidAlgebra.domCongr k k <| AddEquiv.neg G
```

File: Mathlib/Data/Polynomial/Laurent.lean [hunk 2]
```diff
@@ -610,4 +623,31 @@ theorem isLocalization : IsLocalization (Submonoid.closure ({X} : Set R[X])) R[T
 
 end CommSemiring
 
+section Inversion
+
+variable {R : Type _} [CommSemiring R]
```
Comment (eric-wieser) [inline Mathlib/Data/Polynomial/Laurent.lean:628]:
```suggestion
variable {R : Type*} [CommSemiring R]
```

File: Mathlib/LinearAlgebra/Matrix/Charpoly/Coeff.lean [hunk 3]
```diff
@@ -259,3 +260,36 @@ theorem coeff_charpoly_mem_ideal_pow {I : Ideal R} (h : ∀ i j, M i j ∈ I) (k
 #align coeff_charpoly_mem_ideal_pow coeff_charpoly_mem_ideal_pow
 
 end Ideal
+
+section reverse
+
+local notation:9000 R "[T;T⁻¹]" => LaurentPolynomial R
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/Charpoly/Coeff.lean:266]: Can you make this `scoped` notation in the Laurent files? (either in this PR, or a spin-off if you'd prefer)

File: Mathlib/Data/Polynomial/Laurent.lean [hunk 4]
```diff
@@ -610,4 +625,32 @@ theorem isLocalization : IsLocalization (Submonoid.closure ({X} : Set R[X])) R[T
...
+@[simp] lemma invert_T {n : ℤ} : invert (T n : R[T;T⁻¹]) = T (-n) :=
+  AddMonoidAlgebra.domCongr_single _ _ _ _ _
+
+@[simp] lemma invert_apply {f : R[T;T⁻¹]} {n : ℤ} : invert f n = f (-n) := rfl
+
+@[simp] lemma invert_comp_C : invert ∘ (@C R _) = C := by ext; simp
+
+@[simp] lemma invert_C {t : R} : invert (C t) = C t := by ext; simp
+
+lemma involutive_invert : Involutive (invert (R := R)) := fun _ ↦ by ext; simp
```
Comment (eric-wieser) [inline Mathlib/Data/Polynomial/Laurent.lean:644]: Can you add the related lemma that `(invert (R := R)).symm = invert` (which I think is `rfl`?)

File: Mathlib/Data/Polynomial/Reverse.lean [hunk 5]
```diff
@@ -344,6 +347,27 @@ theorem coeff_one_reverse (f : R[X]) : coeff (reverse f) 1 = nextCoeff f := by
...
+  simp [reverse]
+
+@[simp] lemma reverse_mul_X {p : R[X]} : reverse (p * X) = reverse p := by
+  nontriviality R
+  rcases eq_or_ne p 0 with rfl | hp
+  · simp
+  · simp [reverse, hp]
+
+@[simp] lemma reverse_X_mul {p : R[X]} : reverse (X * p) = reverse p := by
+  rw [commute_X p, reverse_mul_X]
```
Comment (eric-wieser) [inline Mathlib/Data/Polynomial/Reverse.lean:361]: Do you think it's worth having `X_pow` versions of these too?

Comment (ocfnash) [reply Mathlib/Data/Polynomial/Reverse.lean:361]: Sure, done.

File: Mathlib/Data/Polynomial/Laurent.lean (hunk 4 above)
Comment (ocfnash) [reply Mathlib/Data/Polynomial/Laurent.lean:644]: Done!

File: Mathlib/Data/Polynomial/Laurent.lean [hunk 6]
```diff
@@ -156,6 +161,9 @@ theorem single_eq_C (r : R) : Finsupp.single 0 r = C r := rfl
 set_option linter.uppercaseLean3 false in
 #align laurent_polynomial.single_eq_C LaurentPolynomial.single_eq_C
 
+@[simp] lemma C_apply {t : R} {n : ℤ} : C t n = if n = 0 then t else 0 := by
+  rw [← single_eq_C, Finsupp.single_apply]; aesop
```
Comment (eric-wieser) [inline Mathlib/Data/Polynomial/Laurent.lean:165]: I don't think this should be a `simp` lemma; [`polynomial.coeff_C`](https://leanprover-community.github.io/mathlib_docs/data/polynomial/basic.html#polynomial.coeff_C) is analogous and also not `simp`.

Comment (eric-wieser) [reply Mathlib/Data/Polynomial/Laurent.lean:165]: Oh nevermind, I guess this is consistent with `MvPolynomial`.

File: Mathlib/Data/Polynomial/Laurent.lean [hunk 7]
```diff
@@ -166,6 +174,9 @@ def T (n : ℤ) : R[T;T⁻¹] :=
 set_option linter.uppercaseLean3 false in
 #align laurent_polynomial.T LaurentPolynomial.T
 
+@[simp] lemma T_apply {m n : ℤ} : (T n : R[T;T⁻¹]) m = if n = m then 1 else 0 :=
```
Comment (eric-wieser) [inline Mathlib/Data/Polynomial/Laurent.lean:177]:
```suggestion
@[simp] lemma T_apply (m n : ℤ) : (T n : R[T;T⁻¹]) m = if n = m then 1 else 0 :=
```

File: Mathlib/Data/Polynomial/Laurent.lean (hunk 6 above)
Comment (eric-wieser) [reply Mathlib/Data/Polynomial/Laurent.lean:165]:
```suggestion
@[simp] lemma C_apply {t : R} {n : ℤ} : C t n = if n = 0 then t else 0 := by
  rw [← single_eq_C, Finsupp.single_apply]; aesop
```
```suggestion
@[simp] lemma C_apply (t : R) (n : ℤ) : C t n = if n = 0 then t else 0 := by
  rw [← single_eq_C, Finsupp.single_apply]; aesop
```

File: Mathlib/Data/Polynomial/Laurent.lean [hunk 8]
```diff
@@ -220,6 +231,10 @@ theorem _root_.Polynomial.toLaurent_C (r : R) : toLaurent (Polynomial.C r) = C r
 set_option linter.uppercaseLean3 false in
 #align polynomial.to_laurent_C Polynomial.toLaurent_C
 
+@[simp]
+theorem _root_.Polynomial.toLaurent_comp_C : toLaurent (R := R) ∘ Polynomial.C = C := by
+  ext; simp
```
Comment (eric-wieser) [inline Mathlib/Data/Polynomial/Laurent.lean:236]: I think it's nice to make it obvious where the interesting version was actually proved here:
```suggestion
theorem _root_.Polynomial.toLaurent_comp_C : toLaurent (R := R) ∘ Polynomial.C = C  :=
  funext Polynomial.toLaurent_C
```

File: Mathlib/Data/Polynomial/Laurent.lean [hunk 9]
```diff
@@ -610,4 +625,34 @@ theorem isLocalization : IsLocalization (Submonoid.closure ({X} : Set R[X])) R[T
...
 
+section Inversion
+
+variable {R : Type*} [CommSemiring R]
+
+/-- The map which substitutes `T ↦ T⁻¹` into a Laurent polynomial. -/
+def invert : R[T;T⁻¹] ≃ₐ[R] R[T;T⁻¹] := AddMonoidAlgebra.domCongr R R <| AddEquiv.neg _
+
+@[simp] lemma invert_T {n : ℤ} : invert (T n : R[T;T⁻¹]) = T (-n) :=
+  AddMonoidAlgebra.domCongr_single _ _ _ _ _
```
Comment (eric-wieser) [inline Mathlib/Data/Polynomial/Laurent.lean:636]:
```suggestion
@[simp] lemma invert_T {n : ℤ} : invert (T n : R[T;T⁻¹]) = T (-n) :=
  AddMonoidAlgebra.domCongr_single _ _ _ _ _
```

File: Mathlib/Data/Polynomial/Laurent.lean [hunk 10]
```diff
@@ -610,4 +625,34 @@ theorem isLocalization : IsLocalization (Submonoid.closure ({X} : Set R[X])) R[T
...
+
+@[simp] lemma invert_apply {f : R[T;T⁻¹]} {n : ℤ} : invert f n = f (-n) := rfl
```
Comment (eric-wieser) [inline Mathlib/Data/Polynomial/Laurent.lean:638]:
```suggestion
@[simp] lemma invert_apply (f : R[T;T⁻¹]) (n : ℤ) : invert f n = f (-n) := rfl
```

File: Mathlib/Data/Polynomial/Laurent.lean [hunk 11]
```diff
@@ -610,4 +625,34 @@ theorem isLocalization : IsLocalization (Submonoid.closure ({X} : Set R[X])) R[T
...
+
+@[simp] lemma invert_comp_C : invert ∘ (@C R _) = C := by ext; simp
+
+@[simp] lemma invert_C {t : R} : invert (C t) = C t := by ext; simp
```
Comment (eric-wieser) [inline Mathlib/Data/Polynomial/Laurent.lean:642]:
```suggestion
@[simp] lemma invert_C (t : R) : invert (C t) = C t := by ext; simp
```

File: Mathlib/Data/Polynomial/Laurent.lean [hunk 12]
```diff
@@ -610,4 +625,34 @@ theorem isLocalization : IsLocalization (Submonoid.closure ({X} : Set R[X])) R[T
...
+
+lemma involutive_invert : Involutive (invert (R := R)) := fun _ ↦ by ext; simp
+
+@[simp] lemma invert_symm : (invert (R := R)).symm = invert := rfl
+
+lemma toLaurent_reverse {p : R[X]} :
```
Comment (eric-wieser) [inline Mathlib/Data/Polynomial/Laurent.lean:648]:
```suggestion
lemma toLaurent_reverse (p : R[X]) :
```

**Review by eric-wieser — COMMENTED** [review]
I've made a start via suggestions, but I think a lot of the arguments to these lemmas should be explicit.

File: Mathlib/Data/Polynomial/Laurent.lean (hunk 9 above)
Comment (ocfnash) [reply Mathlib/Data/Polynomial/Laurent.lean:636]: The suggestion is identical to the original here but I'm assuming you intended to suggest turning `{n : ℤ}` into `(n : ℤ)`.

Comment (eric-wieser) [reply Mathlib/Data/Polynomial/Laurent.lean:636]: Whoops, indeed I was.

File: Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean [hunk 13]
```diff
@@ -92,6 +92,10 @@ def Matrix.charpoly (M : Matrix n n R) : R[X] :=
   (charmatrix M).det
 #align matrix.charpoly Matrix.charpoly
 
+lemma Matrix.charpoly_def (M : Matrix n n R) :
+    M.charpoly = (scalar n (X : R[X]) - (C : R →+* R[X]).mapMatrix M).det :=
+  rfl
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean:97]: I'm not convinced by this lemma vs `rw [char_poly, charmatrix]` at the caller.

File: Mathlib/LinearAlgebra/Matrix/Charpoly/Coeff.lean [hunk 14]
```diff
@@ -259,3 +260,35 @@ theorem coeff_charpoly_mem_ideal_pow {I : Ideal R} (h : ∀ i j, M i j ∈ I) (k
...
+
+open Polynomial
+open LaurentPolynomial hiding C
+
+/-- The right hand side of the equality in this lemma statement is sometimes called the
+"characteristic power series" of a matrix.
+
+It has some advantages over the characteristic polynomial, including the fact that it can be
+extended to infinite dimensions (for appropriate operators). -/
+lemma Matrix.reverse_charpoly (M : Matrix n n R) :
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/Charpoly/Coeff.lean:274]: Is this module-docstring-worthy?

**Review by eric-wieser — APPROVED** [review]
bors d+

Thanks!

**Comment by ocfnash** [comment]
bors merge

**Comment by ocfnash** [comment]
bors merge

**Comment by ocfnash** [comment]
I think our CI is currently flaky and nothing is queued so I'm trying yet again.

bors merge

**Comment by ocfnash** [comment]
bors merge

**Pull request closed without merging**